### DISCUSSION TITLE
apollo-client: bind config to java object

### DISCUSSION THREAD

**Discussion by Anilople** [apolloconfig/apollo#4639] [Ideas]
In spring boot, we can bind config to java object by `@ConfigurationProperties`,

for example
```java
@Configuration
@ConfigurationProperties(prefix = "mail")
public class ConfigProperties {
    
    private String hostName;
    private int port;
    private String from;

    // standard getters and setters
}
```
will bind properties
```properties
#Simple properties
mail.hostname=host@mail.com
mail.port=9000
mail.from=mailer@mail.com
```
to class `ConfigProperties`'s instance.

If we  design some annotation about it, let user can use feature like 
* bind properties to a java object or interface's getter method
* add config change listener to observe java object's field
* running without spring, i.e raw jdk can use it too after use `apollo-client`

Is it will improve user's experience?

reference https://www.baeldung.com/configuration-properties-in-spring-boot

**Comment by nobodyiam** [comment]
Generally speaking, adding a custom annotation to support the object binding will enhance the developer experience for non-spring-boot situations. However, I'm not sure whether this is a strong feature as spring boot is used widely now. 
BTW, what does `add config change listener to observe java object's field` mean? I suppose the object binding will invoke setter methods when some config changes?

**Reply by Anilople** [reply]
> object binding will invoke setter methods when some config changes?

Yes.

maybe
```java
@ConfigBind(prefix = "my.datasource")
public class Datasource {
  public void setUrl(String url) {
    // invoke when my.datasource.url is changed
  }
}
```
or
```java
@ConfigBind(prefix = "my.datasource")
public class Datasource {
  @ConfigChangeInvokeMethod("myMethod")
  private String url;
  public void myMethod(ConfigEvent e) {
    // invoke when my.datasource.url is changed
  }
}
```
or

```java
@ConfigBind(prefix = "my.datasource")
public class Datasource {
  private String url;
  @InvokeAfterFieldChanged(fieldName = "url")
  public void myMethod(ConfigEvent e) {
    // invoke when my.datasource.url is changed
  }
}
```

**Reply by nobodyiam** [reply]
Thanks for the explanation.
Now the only issue is whether this is a strong feature or not.
